plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+ * handle_icap_http_header
+ * Description: This is a good place to determine what to do next based on
+ *              modified http headers from response of icap server.
+ */
+static void
+handle_icap_http_header(TransformData *data)
+{
+  // TSDebug(PLUGIN_NAME, "Handling http header");
+  int64_t pos                  = data->http_header.find("\r\n");
+  std::string http_status_line = pos != (int64_t)std::string::npos ? data->http_header.substr(0, pos) : data->http_header;
+  /* find content length from header if any */
+  std::smatch sm;
+  std::regex e("(Content-Length: )([[:digit:]]+)");
+  regex_search(data->http_header, sm, e);
+  if (sm.size()) {
+    data->icap_reply_content_length = std::stoll(sm[2].str().c_str(), NULL, 10);
+  }
+  /* Replace header with the returned header from icap server */
+  TSMBuffer bufp;
+  TSMLoc hdr_loc;
+  TSHttpParser parser;
+  const char *raw_resp = data->http_header.c_str();
+
+  if (TSHttpTxnTransformRespGet(data->txn, &bufp, &hdr_loc) != TS_SUCCESS) {
+    TSError("[%s] Couldn't retrieve transform response header", PLUGIN_NAME);
+    return;
+  }
+  /* Clear all headers from the transform response */
+  if (TSMimeHdrFieldsClear(bufp, hdr_loc) == TS_ERROR) {
+    TSError("[%s] Couldn't clear client response header", PLUGIN_NAME);
+    return;
+  }
+  /* Create the new header using http header in icap response */
+  parser = TSHttpParserCreate();
+  TSHttpHdrParseResp(parser, bufp, hdr_loc, &raw_resp, raw_resp + data->http_header.size());
+
+  TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+}
+
+/*
+ * handle_read_http_body
+ * Description: This function handles reading http body from icap server, as
+ *              well as writing body to downstream. It should be called whenever
+ *              new data is available to read or WRITE_READY is received from
+ *              downstream.
+ */
+static int
+handle_read_http_body(TSCont contp, TransformData *data)
+{
+  int64_t avail = TSIOBufferReaderAvail(data->icap_resp_reader);
+
+  if (avail > 0) {
+    /* Read the chunk length if one is not available */
+    if (data->http_body_chunk_length <= 0) {
+      int64_t data_len;
+      const char *buf;
+      int64_t consumed    = data->chunk_length_str.size();
+      TSIOBufferBlock blk = TSIOBufferReaderStart(data->icap_resp_reader);
+
+      while (blk != nullptr) {
+        buf               = TSIOBufferBlockReadStart(blk, data->icap_resp_reader, &data_len);
+        std::string chunk = std::string(buf, data_len);
+        /* keep the read string in body_length in case complete data haven't arrived */
+        data->chunk_length_str += chunk;
+        /* Look for end of reply token */
+        if (data->chunk_length_str.find("\r\n0\r\n\r\n") == 0) {
+          TSVIONBytesSet(data->output_vio, data->http_body_total_length_written);
+          return 0;
+        }
+        /* Look for hex string indicating chunk length */
+        std::smatch sm;
+        std::regex e("(\r\n)([[:xdigit:]]+)(\r\n)");
```

Comment:
Seems like this should have non-local scope and be constructed in TSPluginInit().